To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Installation

In [ ]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [ ]:
import os, wandb
from huggingface_hub import login

os.environ["WANDB_API_KEY"] = "810a489e4029d94651a3868c08a49812cc1e4cc7"
os.environ["HUGGINGFACE_API_KEY"] = "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"

In [ ]:
hf_token = os.environ["HUGGINGFACE_API_KEY"]
login(hf_token)

secret_key = os.environ["WANDB_API_KEY"]
wandb.login(key=secret_key)

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: sonhh2003 (sonhh2003-hanoi-university-of-science-and-technology) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

In [ ]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Meta-Llama-3.1-8B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.4.7: Fast Llama patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/5.96G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/235 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/50.6k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/459 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [ ]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.4.7 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [ ]:
from datasets import Dataset, load_dataset

SYSTEM_PROMPT = """
Specific Requirements:
  Read and Analyze the Text: Carefully read the entire content of the cookie policy to understand the types of cookies used, their purposes, storage duration, detailed descriptions, and information about the owner (first party or third party).
  Identify Described Cookies: Search for sections, paragraphs, or text that describe specific cookies. Each cookie will usually have its own entry or be listed in a list. If not found cookies, return [] and dont say anymore.

  Extract Detailed Information for Each Cookie: For each identified cookie, carefully extract the following information from the policy:
    +) "cookie_name": Find the technical name or descriptive name of the cookie exactly as it is mentioned in the policy.
    +) "declared_purpose": cookie's purpose. With the following options:
        +) Strictly Necessary: if cookies essential for the website to function correctly with basic features (e.g., managing login sessions, shopping cart functionality, secure navigation).
        +) Functionality: if cookies that help personalize the user experience and provide additional features (e.g., remembering language preferences, page layouts).
        +) Analytical: if cookies used to collect data and understand user behavior on the website (e.g., number of visits, most viewed pages).
        +) Targeting/Advertising/Marketing: if cookies used to display personalized advertisements based on user behavior or interests.
        +) Performance: if cookies that help improve the technical performance of the website (e.g., optimizing page loading, content distribution).
        +) Social Sharing: if cookies that enable sharing content from the website to social media platforms.
        +) Null: if not provide any specific information about the purpose of this cookie.
    +) "declared_retention": duration this cookie will be stored on the user's device (e.g., "6 months", "24 hours", "Session", "Persistent", "1 minute", "Until deleted"). Record this information as accurately as possible according to the original text.
    +) "declared_third_parties": An array of third parties involved in the use of this cookie (if any). ["First Party"] indicates a first-party cookie set by the website itself.
    +) "declared_description": The content taken directly from the text, without fabrication, using exactly what they have described.
"""

def preprocess(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    prompt = [
        {'role': 'system', 'content': SYSTEM_PROMPT},
        {'role': 'user', 'content': content}
    ]

    return {
        'prompt': prompt,
        'answer': example['label']
        # 'answer': '[]'
    }

dataset = load_dataset("sonhask/detect-cookie-with-label")
dataset = dataset['train'].map(preprocess)


In [ ]:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
Your task is to read a cookie policy text and extract detailed information about each cookie mentioned in that policy.
RESPONSE Format: If no cookies are specifically described, only response []. Otherwise, Write a response that is JSON array following structure:
  [
    {{
      "cookie_name": "cookie_name",
      "declared_purpose": "declared_purpose",
      "declared_retention": "declared_retention",
      "declared_third_parties": ["declared_third_parties"],
      "declared_description": "declared_description"
    \}}, ...
  ]
### Instruction:
{}

### Input:
{}

### Response:
{}"""

SYSTEM_PROMPT = """
Specific Requirements:
  Read and Analyze the Text: Carefully read the entire content of the cookie policy to understand the types of cookies used, their purposes, storage duration, detailed descriptions, and information about the owner (first party or third party).
  Identify Described Cookies: Search for sections, paragraphs, or text that describe specific cookies. Each cookie will usually have its own entry or be listed in a list. If not found cookies, return [] and dont say anymore.

  Extract Detailed Information for Each Cookie: For each identified cookie, carefully extract the following information from the policy:
    +) "cookie_name": Find the technical name or descriptive name of the cookie exactly as it is mentioned in the policy.
    +) "declared_purpose": cookie's purpose. With the following options:
        +) Strictly Necessary: if cookies essential for the website to function correctly with basic features (e.g., managing login sessions, shopping cart functionality, secure navigation).
        +) Functionality: if cookies that help personalize the user experience and provide additional features (e.g., remembering language preferences, page layouts).
        +) Analytical: if cookies used to collect data and understand user behavior on the website (e.g., number of visits, most viewed pages).
        +) Targeting/Advertising/Marketing: if cookies used to display personalized advertisements based on user behavior or interests.
        +) Performance: if cookies that help improve the technical performance of the website (e.g., optimizing page loading, content distribution).
        +) Social Sharing: if cookies that enable sharing content from the website to social media platforms.
        +) Null: if not provide any specific information about the purpose of this cookie.
    +) "declared_retention": duration this cookie will be stored on the user's device (e.g., "6 months", "24 hours", "Session", "Persistent", "1 minute", "Until deleted"). Record this information as accurately as possible according to the original text.
    +) "declared_third_parties": An array of third parties involved in the use of this cookie (if any). ["First Party"] indicates a first-party cookie set by the website itself.
    +) "declared_description": The content taken directly from the text, without fabrication, using exactly what they have described.
"""

EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    instruction = SYSTEM_PROMPT
    input       = content
    output      = str(example['label'])


    text = alpaca_prompt.format(instruction, input, output) + EOS_TOKEN
    return { "text" : text, }

from datasets import load_dataset
dataset = load_dataset("sonhask/detect-cookie-with-label", split = "train")
dataset = dataset.map(formatting_prompts_func)

Map:   0%|          | 0/12441 [00:00<?, ? examples/s]

In [ ]:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
Your task is to read a cookie policy text and extract detailed information about each cookie mentioned in that policy.
RESPONSE Format: If no cookies are specifically described, only response []. Otherwise, Write a response that is JSON array following structure:
  [
    {{
      "cookie_name": "cookie_name",
      "declared_purpose": "declared_purpose",
      "declared_retention": "declared_retention",
      "declared_third_parties": ["declared_third_parties"],
      "declared_description": "declared_description"
    }}, ...
  ]
### Instruction:
{}

### Input:
{}

### Response:
{}"""

SYSTEM_PROMPT = """
Specific Requirements:
  Read and Analyze the Text: Carefully read the entire content of the cookie policy to understand the types of cookies used, their purposes, storage duration, detailed descriptions, and information about the owner (first party or third party).
  Identify Described Cookies: Search for sections, paragraphs, or text that describe specific cookies. Each cookie will usually have its own entry or be listed in a list. If not found cookies, return [] and dont say anymore.

  Extract Detailed Information for Each Cookie: For each identified cookie, carefully extract the following information from the policy:
    +) "cookie_name": Find the technical name or descriptive name of the cookie exactly as it is mentioned in the policy.
    +) "declared_purpose": cookie's purpose. With the following options:
        +) Strictly Necessary: if cookies essential for the website to function correctly with basic features (e.g., managing login sessions, shopping cart functionality, secure navigation).
        +) Functionality: if cookies that help personalize the user experience and provide additional features (e.g., remembering language preferences, page layouts).
        +) Analytical: if cookies used to collect data and understand user behavior on the website (e.g., number of visits, most viewed pages).
        +) Targeting/Advertising/Marketing: if cookies used to display personalized advertisements based on user behavior or interests.
        +) Performance: if cookies that help improve the technical performance of the website (e.g., optimizing page loading, content distribution).
        +) Social Sharing: if cookies that enable sharing content from the website to social media platforms.
        +) Null: if not provide any specific information about the purpose of this cookie.
    +) "declared_retention": duration this cookie will be stored on the user's device (e.g., "6 months", "24 hours", "Session", "Persistent", "1 minute", "Until deleted"). Record this information as accurately as possible according to the original text.
    +) "declared_third_parties": An array of third parties involved in the use of this cookie (if any). ["First Party"] indicates a first-party cookie set by the website itself.
    +) "declared_description": The content taken directly from the text, without fabrication, using exactly what they have described.
"""

EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    instruction = SYSTEM_PROMPT
    input       = content
    output      = example['label']
    text = alpaca_prompt.format(instruction, input, output) + EOS_TOKEN
    return { "text" : text}

from datasets import load_dataset
dataset = load_dataset("sonhask/detect-cookie-with-label", split = "train")


In [ ]:
dataset = dataset.filter(lambda example: example['english_content'] != None)
dataset

Filter:   0%|          | 0/12441 [00:00<?, ? examples/s]

Dataset({
    features: ['url', 'lang', 'original_text', 'tables', 'processed_text', 'english_content', 'english_table', 'translated_tables', 'label', 'Unnamed: 0'],
    num_rows: 11790
})

In [ ]:
def compute_token_length(example):
    length = (len(SYSTEM_PROMPT) + len(example['english_content']) + len(example['english_table']) + len(example['label']))*0.331 + 50
    return length

dataset = dataset.filter(lambda example: compute_token_length(example) <= 2048)
len(dataset)

Filter:   0%|          | 0/10153 [00:00<?, ? examples/s]

7740

In [ ]:
dataset = dataset.map(formatting_prompts_func)

Map:   0%|          | 0/7740 [00:00<?, ? examples/s]

In [ ]:
def thong_ke(dataset):
    empty_label_count = 0
    non_empty_label_count = 0

    for i in dataset:
        if i['label'] == '[]':
            empty_label_count += 1
        else:
            non_empty_label_count += 1

    print("#[]: ", empty_label_count)
    print("#Has cookies: ", non_empty_label_count)

thong_ke(dataset)

#[]:  6836
#Has cookies:  904


In [ ]:
def tokenize_batch(examples):
    toks = tokenizer(
        examples["text"],
        padding="max_length",
        truncation=True,
        max_length=max_seq_length,
    )
    # Copy input_ids as labels for causal LM
    toks["labels"] = toks["input_ids"].copy()
    return toks

dataset = dataset.map(tokenize_batch, batched=True)


Map:   0%|          | 0/7740 [00:00<?, ? examples/s]

In [ ]:
dataset = dataset.remove_columns(['url', 'lang', 'original_text', 'tables', 'processed_text', 'english_content', 'english_table', 'translated_tables', 'label', 'Unnamed: 0', 'text', ])

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [ ]:
run = wandb.init(
    project="Fine-tune Llama 3_1 8B GRPO",
)

In [ ]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    processing_class = tokenizer,
    train_dataset = dataset,
    # dataset_text_field = "text",
    max_seq_length = max_seq_length,
    # dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        num_train_epochs = 1, # Set this for 1 full training run.
        # max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit", # "paged_adamw_8bit"
        weight_decay = 0.01,
        lr_scheduler_type = "cosine",
        seed = 3407,
        output_dir = "/content/drive/MyDrive/Llama3.1_(8B)_SFT",
        report_to = "wandb", # Use this for WandB etc
        save_strategy="steps",                # Hoặc "epoch" nếu bạn muốn lưu sau mỗi epoch
        save_steps=25,                       # Lưu mỗi 50 steps (tùy bạn chỉnh)
        save_total_limit=3,                   # Giới hạn số lượng checkpoint (xóa cũ)
    ),
)

In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
7.135 GB of memory reserved.


In [ ]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 7,740 | Num Epochs = 1 | Total steps = 967
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 41,943,040/8,000,000,000 (0.52% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

462.7198 seconds used for training.
7.71 minutes used for training.
Peak reserved memory = 7.922 GB.
Peak reserved memory for training = 1.938 GB.
Peak reserved memory % of max memory = 53.716 %.
Peak reserved memory for training % of max memory = 13.141 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [ ]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[
    alpaca_prompt.format(
        "Continue the fibonnaci sequence.", # instruction
        "1, 1, 2, 3, 5, 8", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

outputs = model.generate(**inputs, max_new_tokens = 64, use_cache = True)
tokenizer.batch_decode(outputs)

['<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nContinue the fibonnaci sequence.\n\n### Input:\n1, 1, 2, 3, 5, 8\n\n### Response:\n13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025']

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[
    alpaca_prompt.format(
        "Continue the fibonnaci sequence.", # instruction
        "1, 1, 2, 3, 5, 8", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Continue the fibonnaci sequence.

### Input:
1, 1, 2, 3, 5, 8

### Response:
13, 21, 34, 55, 89, 144<|end_of_text|>


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# alpaca_prompt = You MUST copy from above!

inputs = tokenizer(
[
    alpaca_prompt.format(
        "What is a famous tall tower in Paris?", # instruction
        "", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
What is a famous tall tower in Paris?

### Input:


### Response:
One of the most famous and iconic tall towers in Paris is the Eiffel Tower. Standing at 324 meters (1,063 feet) tall, this wrought iron tower is a symbol of the city and a must-see attraction for tourists from all over the world.<|end_of_text|>


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
